### ISSUE TITLE
DataSource in CakePHP 2.8.0

### ISSUE THREAD

**Issue by kicaj** [cakephp/cakephp#8281]
At line 371: https://github.com/cakephp/cakephp/blob/2.x/lib/Cake/Model/Datasource/DataSource.php#L371
sometimes I get error like this: 
`Illegal offset type in isset or empty [APP\Vendor\cakephp\cakephp\lib\Cake\Model\Datasource\DataSource.php, line 371]`

Problem is when $assocData from one line before is empty. 

I think `if` statement at line 371 shoud look like this:
`if (!empty($assocData) && isset($data[$assocData]) && isset($data[$assocData][$insertKey])) {`

**Comment by markstory** [comment]
How can someone reproduce the issue you're having?

**Comment by kicaj** [comment]
Hmm, problem is when I work with Oracle driver (https://bitbucket.org/gregnuj/cakephp-2.x-oracle-datasource.git), with my a little modification to work with 2.8.

When line 371 show error, my data look like this:

$stack: 
```
Array
(
    [0] => ProductCategory
    [_joined] => Array
        (
        )
)
```
array_reverse($stack):
```
Array
(
    [_joined] => Array
        (
        )
    [0] => ProductCategory
)
```

My relation is Product HABTM ProductCategory

I thinks the line 371, should look like:
`if (is_string($assocData) && !empty($assocData) && isset($data[$assocData]) && isset($data[$assocData][$insertKey])) {`

**Comment by markstory** [comment]
Fixed in 7a59070
